pkg/controller/syncmachineset/syncmachineset_controller.go

```diff
@@ -0,0 +1,555 @@
+	remoteMachineSet, err := remoteClusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterAPINamespace).Get(ms.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		msLog.Infof("creating new machineSet API object in remote cluster")
+
+		// Lookup clusterVersion for machineSet
+		coVersionNamespace, coVersionName := ms.Spec.ClusterVersionRef.Namespace, ms.Spec.ClusterVersionRef.Name
+		coClusterVersion, err := c.client.ClusteroperatorV1alpha1().ClusterVersions(coVersionNamespace).Get(coVersionName, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("cannot retrieve cluster version %s/%s: %v", coVersionNamespace, coVersionName, err)
+		}
+
+		remoteMachineSet = buildClusterAPIMachineSet(ms, coClusterVersion)
+		_, erree := remoteClusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterAPINamespace).Create(remoteMachineSet)
+		if erree != nil {
+			return fmt.Errorf("error creating machineSet API object in remote cluster: %v", erree)
+		}
+	} else if err != nil {
+		return err
+	} else {
+		// TODO: patch/update here?
+		msLog.Debugf("machineSet API object already exists in remote cluster")
+	}
+
+	return nil
+}
+
+func buildClusterAPIMachineSet(machineSet *cov1.MachineSet, clusterVersion *cov1.ClusterVersion) *clusterapiv1.MachineSet {
+	if machineSet.Annotations == nil {
+		machineSet.Annotations = map[string]string{}
+	}
+	machineSet.Annotations["cluster-operator.openshift.io/cluster-version"] = serializeCOResource(clusterVersion)
+
+	capiMachineSet := clusterapiv1.MachineSet{}
+	capiMachineSet.Name = machineSet.Name
+	capiMachineSet.Namespace = remoteClusterAPINamespace
+
+	// Set replicas = 0 so that we don't create machines yet.
+	// TODO: remove me when ready to create machines from remote.
+	// replicas := int32(machineSet.Spec.Size)
+	replicas := int32(0)
+	capiMachineSet.Spec.Replicas = &replicas
+
+	capiMachineSet.Spec.Selector.MatchLabels = map[string]string{"machineset": machineSet.Name}
+
+	machineTemplate := clusterapiv1.MachineTemplateSpec{}
+	machineTemplate.Labels = map[string]string{"machineset": machineSet.Name}
+	machineTemplate.Spec.Labels = map[string]string{"machineset": machineSet.Name}
+	machineTemplate.Spec.ProviderConfig.Value = &runtime.RawExtension{
+		Raw: []byte(serializeCOResource(machineSet)),
+	}
+	capiMachineSet.Spec.Template = machineTemplate
+
+	return &capiMachineSet
+}
+
+func buildClusterAPICluster(cluster *cov1.Cluster) *clusterapiv1.Cluster {
+	capiCluster := clusterapiv1.Cluster{}
+	capiCluster.Name = cluster.Name
+	capiCluster.Spec.ProviderConfig.Value = &runtime.RawExtension{
+		Raw: []byte(serializeCOResource(cluster)),
+	}
+
+	// These are unused, dummy values.
+	capiCluster.Spec.ClusterNetwork.ServiceDomain = "cluster-api.k8s.io"
+	capiCluster.Spec.ClusterNetwork.Pods.CIDRBlocks = []string{"10.10.0.0/16"}
+	capiCluster.Spec.ClusterNetwork.Services.CIDRBlocks = []string{"172.30.0.0/16"}
+
+	return &capiCluster
+}
+
+func serializeCOResource(object runtime.Object) string {
```

Comment:
Return a `[]byte` rather than a `string` as we ultimately want a `[]byte` anyway. This saves us from converting the `[]byte]` to a `string` and then immediately back to a `[]byte`. Call `Bytes()` instead of `String()` on the `bytes.Buffer` to get a `[]byte` to return.